File: core/chunking-manifest/js/src/ledger.mjs
import {
  closeSync, mkdtempSync, openSync, readSync, rmSync, writeSync,
} from 'node:fs';
import { tmpdir } from 'node:os';
import { join, resolve } from 'node:path';
import { ObjectRef } from '@opengamevcs/object-model';
import { fail, wrap } from './errors.mjs';

export const LEDGER_RECORD_BYTES = 48;
export const LEDGER_MAXIMUM_BYTES = 64 * 1024 * 1024;
export const LEDGER_DEFAULTS = Object.freeze({
  maxMemoryBytes: 1_048_576,
  maxScratchBytes: 64 * 1024 * 1024,
});

function limit(value, fallback) {
  const selected = value ?? fallback;
  if (!Number.isSafeInteger(selected) || selected < 0 || selected > LEDGER_MAXIMUM_BYTES) {
    fail('CHUNK_RESOURCE_INVALID');
  }
  return selected;
}

function encode(record) {
  if (!(record.digest instanceof Uint8Array) || record.digest.byteLength !== 32 ||
      !Number.isSafeInteger(record.length) || record.length < 0 ||
      !Number.isSafeInteger(record.boundary) || record.boundary < 0) {
    fail('CHUNK_RESOURCE_INVALID');
  }
  const bytes = Buffer.allocUnsafe(LEDGER_RECORD_BYTES);
  Buffer.from(record.digest.buffer, record.digest.byteOffset, record.digest.byteLength).copy(bytes, 0);
  bytes.writeBigUInt64BE(BigInt(record.length), 32);
  bytes.writeBigUInt64BE(BigInt(record.boundary), 40);
  return bytes;
}

function decode(bytes) {
  const digest = Buffer.from(bytes.subarray(0, 32));
  const length = Number(bytes.readBigUInt64BE(32));
  const boundary = Number(bytes.readBigUInt64BE(40));
  const reference = new ObjectRef(1, digest);
  return Object.freeze({ digest, length, boundary, objectId: reference.toString(), reference });
}

/**
 * Append-only, repeatable fixed-record ledger. It uses at most maxMemoryBytes
 * for records, then atomically switches to an owner-private scratch file.
 * dispose() is idempotent and removes every scratch artifact.
 */
export function createLedger(options = {}) {
  const maxMemoryBytes = limit(options.maxMemoryBytes, LEDGER_DEFAULTS.maxMemoryBytes);
  const maxScratchBytes = limit(options.maxScratchBytes, LEDGER_DEFAULTS.maxScratchBytes);
  const memoryRecordMaximum = Math.floor(maxMemoryBytes / LEDGER_RECORD_BYTES);
  const scratchRoot = resolve(options.scratchDirectory ?? tmpdir());
  let resident = Buffer.alloc(0);
  let residentCount = 0;
  let directory;
  let path;
  let descriptor;
  let count = 0;
  let scratchBytes = 0;
  let peakMemoryBytes = 0;
  let peakScratchBytes = 0;
  let disposed = false;

  function assertOpen() {
    if (disposed) fail('CHUNK_SESSION_FINISHED');
  }

  function writeRecord(bytes) {
    if (scratchBytes > maxScratchBytes - LEDGER_RECORD_BYTES) {
      fail('CHUNK_SCRATCH_EXHAUSTED', { maxScratchBytes });
    }
    let offset = 0;
    while (offset < bytes.length) {
      const written = writeSync(descriptor, bytes, offset, bytes.length - offset, scratchBytes + offset);
      if (written <= 0) fail('CHUNK_SCRATCH_EXHAUSTED', { maxScratchBytes });
      offset += written;
    }
    scratchBytes += bytes.length;
    peakScratchBytes = Math.max(peakScratchBytes, scratchBytes);
  }

  function spill() {
    if (descriptor !== undefined) return;
    if (maxScratchBytes < (residentCount + 1) * LEDGER_RECORD_BYTES) {
      fail('CHUNK_SCRATCH_EXHAUSTED', { maxScratchBytes });
    }
    try {
      directory = mkdtempSync(join(scratchRoot, 'ogvcs-chunk-ledger-'));
      path = join(directory, 'records.bin');
      descriptor = openSync(path, 'wx+', 0o600);
      for (let index = 0; index < residentCount; index += 1) {
        writeRecord(resident.subarray(index * LEDGER_RECORD_BYTES, (index + 1) * LEDGER_RECORD_BYTES));
      }
      resident = Buffer.alloc(0);
      residentCount = 0;
    } catch (cause) {
      try { if (descriptor !== undefined) closeSync(descriptor); } catch {}
      descriptor = undefined;
      if (directory !== undefined) {
        try { rmSync(directory, { recursive: true, force: true }); } catch {}
      }
      directory = undefined;
      path = undefined;
      if (cause?.code === 'CHUNK_SCRATCH_EXHAUSTED') throw cause;
      throw wrap('CHUNK_SCRATCH_EXHAUSTED', cause, { maxScratchBytes });
    }
  }

  function append(record) {
    assertOpen();
    const bytes = encode(record);
    if (descriptor === undefined && residentCount < memoryRecordMaximum) {
      if (resident.byteLength === 0) {
        resident = Buffer.allocUnsafe(memoryRecordMaximum * LEDGER_RECORD_BYTES);
        peakMemoryBytes = resident.byteLength;
      }
      bytes.copy(resident, residentCount * LEDGER_RECORD_BYTES);
      residentCount += 1;
    }
    else {
      spill();
      writeRecord(bytes);
    }
    count += 1;
    return count;
  }

  function *records() {
    assertOpen();
    if (descriptor === undefined) {
      for (let index = 0; index < residentCount; index += 1) {
        yield decode(resident.subarray(index * LEDGER_RECORD_BYTES, (index + 1) * LEDGER_RECORD_BYTES));
      }
      return;
    }
    const bytes = Buffer.allocUnsafe(LEDGER_RECORD_BYTES);
    for (let index = 0; index < count; index += 1) {
      let offset = 0;
      while (offset < bytes.length) {
        const read = readSync(descriptor, bytes, offset, bytes.length - offset,
          index * LEDGER_RECORD_BYTES + offset);
        if (read <= 0) fail('CHUNK_SESSION_FAILED');
        offset += read;
      }
      yield decode(bytes);
    }
  }

  function metrics() {
    return Object.freeze({
      records: count,
      memoryBytes: descriptor === undefined ? residentCount * LEDGER_RECORD_BYTES : 0,
      peakMemoryBytes,
      scratchBytes,
      peakScratchBytes,
      spilled: descriptor !== undefined,
    });
  }

  function dispose() {
    if (disposed) return;
    disposed = true;
    resident = Buffer.alloc(0);
    residentCount = 0;
    if (descriptor !== undefined) {
      try { closeSync(descriptor); } catch {}
      descriptor = undefined;
    }
    if (directory !== undefined) {
      try { rmSync(directory, { recursive: true, force: true }); } catch {}
      directory = undefined;
      path = undefined;
    }
  }

  return Object.freeze({ append, dispose, get count() { return count; }, metrics, records });
}

File: core/chunking-manifest/rust/src/ledger.rs
use std::{
    fs::{create_dir, remove_dir, remove_file, File, OpenOptions},
    io::{Read, Seek, SeekFrom, Write},
    path::{Path, PathBuf},
    sync::atomic::{AtomicU64, Ordering},
};

use crate::ChunkError;

pub const LEDGER_RECORD_BYTES: u64 = 48;
pub const DEFAULT_LEDGER_MEMORY_BYTES: u64 = 1_048_576;
pub const DEFAULT_SCRATCH_BYTES: u64 = 64 * 1024 * 1024;
pub const LEDGER_MAXIMUM_BYTES: u64 = 64 * 1024 * 1024;

#[derive(Clone, Debug)]
pub struct LedgerOptions {
    pub max_memory_bytes: u64,
    pub max_scratch_bytes: u64,
    pub scratch_directory: PathBuf,
}

impl Default for LedgerOptions {
    fn default() -> Self {
        Self {
            max_memory_bytes: DEFAULT_LEDGER_MEMORY_BYTES,
            max_scratch_bytes: DEFAULT_SCRATCH_BYTES,
            scratch_directory: std::env::temp_dir(),
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct LedgerMetrics {
    pub records: u64,
    pub memory_bytes: u64,
    pub peak_memory_bytes: u64,
    pub scratch_bytes: u64,
    pub peak_scratch_bytes: u64,
    pub spilled: bool,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) struct LedgerRecord {
    pub digest: [u8; 32],
    pub length: u64,
    pub boundary: u64,
}

impl LedgerRecord {
    fn encode(self) -> [u8; LEDGER_RECORD_BYTES as usize] {
        let mut bytes = [0u8; LEDGER_RECORD_BYTES as usize];
        bytes[..32].copy_from_slice(&self.digest);
        bytes[32..40].copy_from_slice(&self.length.to_be_bytes());
        bytes[40..48].copy_from_slice(&self.boundary.to_be_bytes());
        bytes
    }

    fn decode(bytes: [u8; LEDGER_RECORD_BYTES as usize]) -> Self {
        let mut digest = [0u8; 32];
        digest.copy_from_slice(&bytes[..32]);
        Self {
            digest,
            length: u64::from_be_bytes(bytes[32..40].try_into().expect("eight bytes")),
            boundary: u64::from_be_bytes(bytes[40..48].try_into().expect("eight bytes")),
        }
    }
}

pub(crate) struct Ledger {
    options: LedgerOptions,
    resident: Vec<[u8; LEDGER_RECORD_BYTES as usize]>,
    file: Option<File>,
    directory: Option<PathBuf>,
    path: Option<PathBuf>,
    count: u64,
    scratch_bytes: u64,
    peak_memory_bytes: u64,
    peak_scratch_bytes: u64,
}

impl Ledger {
    pub(crate) fn new(options: LedgerOptions) -> Result<Self, ChunkError> {
        if options.max_memory_bytes > LEDGER_MAXIMUM_BYTES
            || options.max_scratch_bytes > LEDGER_MAXIMUM_BYTES
            || !options.scratch_directory.is_dir()
        {
            return Err(ChunkError::ResourceInvalid);
        }
        let capacity = usize::try_from(
            (options.max_memory_bytes / LEDGER_RECORD_BYTES).min(1_048_576),
        )
        .map_err(|_| ChunkError::ResourceInvalid)?;
        let peak_memory_bytes = capacity as u64 * LEDGER_RECORD_BYTES;
        Ok(Self {
            options,
            resident: Vec::with_capacity(capacity),
            file: None,
            directory: None,
            path: None,
            count: 0,
            scratch_bytes: 0,
            peak_memory_bytes,
            peak_scratch_bytes: 0,
        })
    }

    pub(crate) const fn len(&self) -> u64 {
        self.count
    }

    fn private_directory(root: &Path) -> Result<PathBuf, ChunkError> {
        static NEXT: AtomicU64 = AtomicU64::new(0);
        for _ in 0..1024 {
            let nonce = NEXT.fetch_add(1, Ordering::Relaxed);
            let path = root.join(format!(
                "ogvcs-chunk-ledger-{}-{nonce}",
                std::process::id()
            ));
            match create_dir(&path) {
                Ok(()) => return Ok(path),
                Err(error) if error.kind() == std::io::ErrorKind::AlreadyExists => continue,
                Err(_) => return Err(ChunkError::ScratchExhausted),
            }
        }
        Err(ChunkError::ScratchExhausted)
    }

    fn write_record(&mut self, bytes: &[u8; LEDGER_RECORD_BYTES as usize]) -> Result<(), ChunkError> {
        let next = self
            .scratch_bytes
            .checked_add(LEDGER_RECORD_BYTES)
            .ok_or(ChunkError::ScratchExhausted)?;
        if next > self.options.max_scratch_bytes {
            return Err(ChunkError::ScratchExhausted);
        }
        self.file
            .as_mut()
            .ok_or(ChunkError::SessionFailed)?
            .write_all(bytes)
            .map_err(|_| ChunkError::ScratchExhausted)?;
        self.scratch_bytes = next;
        self.peak_scratch_bytes = self.peak_scratch_bytes.max(next);
        Ok(())
    }

    fn spill(&mut self) -> Result<(), ChunkError> {
        if self.file.is_some() {
            return Ok(());
        }
        let required = (self.resident.len() as u64 + 1)
            .checked_mul(LEDGER_RECORD_BYTES)
            .ok_or(ChunkError::ScratchExhausted)?;
        if required > self.options.max_scratch_bytes {
            return Err(ChunkError::ScratchExhausted);
        }
        let directory = Self::private_directory(&self.options.scratch_directory)?;
        let path = directory.join("records.bin");
        let file = match OpenOptions::new().read(true).write(true).create_new(true).open(&path) {
            Ok(file) => file,
            Err(_) => {
                let _ = remove_dir(&directory);
                return Err(ChunkError::ScratchExhausted);
            }
        };
        self.directory = Some(directory);
        self.path = Some(path);
        self.file = Some(file);
        let resident = std::mem::take(&mut self.resident);
        for record in resident {
            if let Err(error) = self.write_record(&record) {
                self.cleanup();
                return Err(error);
            }
        }
        Ok(())
    }

    pub(crate) fn append(&mut self, record: LedgerRecord) -> Result<(), ChunkError> {
        let resident_maximum = self.options.max_memory_bytes / LEDGER_RECORD_BYTES;
        let bytes = record.encode();
        if self.file.is_none() && (self.resident.len() as u64) < resident_maximum {
            self.resident.push(bytes);
        } else {
            self.spill()?;
            self.write_record(&bytes)?;
        }
        self.count = self.count.checked_add(1).ok_or(ChunkError::CountExceeded)?;
        Ok(())
    }

    pub(crate) fn for_each(
        &mut self,
        mut consume: impl FnMut(LedgerRecord) -> Result<(), ChunkError>,
    ) -> Result<(), ChunkError> {
        if let Some(file) = self.file.as_mut() {
            file.seek(SeekFrom::Start(0))
                .map_err(|_| ChunkError::SessionFailed)?;
            for _ in 0..self.count {
                let mut bytes = [0u8; LEDGER_RECORD_BYTES as usize];
                file.read_exact(&mut bytes)
                    .map_err(|_| ChunkError::SessionFailed)?;
                consume(LedgerRecord::decode(bytes))?;
            }
        } else {
            for bytes in &self.resident {
                consume(LedgerRecord::decode(*bytes))?;
            }
        }
        Ok(())
    }

    pub(crate) fn metrics(&self) -> LedgerMetrics {
        LedgerMetrics {
            records: self.count,
            memory_bytes: if self.file.is_none() {
                self.resident.len() as u64 * LEDGER_RECORD_BYTES
            } else {
                0
            },
            peak_memory_bytes: self.peak_memory_bytes,
            scratch_bytes: self.scratch_bytes,
            peak_scratch_bytes: self.peak_scratch_bytes,
            spilled: self.file.is_some(),
        }
    }

    pub(crate) fn cleanup(&mut self) {
        self.file.take();
        if let Some(path) = self.path.take() {
            let _ = remove_file(path);
        }
        if let Some(directory) = self.directory.take() {
            let _ = remove_dir(directory);
        }
        self.resident = Vec::new();
    }
}

impl Drop for Ledger {
    fn drop(&mut self) {
        self.cleanup();
    }
}
